### PR TITLE
Starshapes clarity kit

### PR DESCRIPTION
**Pull request by lguelzow**
Changes to models.py:

added 2 atmosphere models for experimental sites in China for GRAND
removed a deprecated module that didn't run for me (xrange -> range)
slightly widened a range in an atmospheric height calculation because it crashed for the 50-200 MHz frequency band

Changes to coordinatesystems.py:

just a typo in a print command

Changes to generate_coreas_sim.py:

extensively expanded the "write_list_star_pattern" function with detailed documentation. Everything new is at least briefly described. The default output of the function stays the same.
The main change is to replace the undocumented rotation by -90 degrees to transform to the Auger coordinate system with an explicit rotation matrix that is well documented. It also enables generating starshapes in Corsika's native coordinate system with a new input option: Auger_CS. By default it is set to True and performs the same -90 degree rotation as the current version of the code.
Error messages are included to catch inputs in the wrong units for angle and distances.
The function also now returns the azimuth angle that has to be written in the Corsika input to get the right shower direction for the corresponding starshape.
Added another optional input option to the function that writes a second .list file containing antenna positions in the shower plane coordinate system to enable visual checks of the starshape.

Added new functions to generate pre-made antenna rings with variable density that can conform to the shower geometry.

### CODE REVIEW

File: radiotools/coreas/generate_coreas_sim.py [hunk 1]
```diff
@@ -171,67 +171,233 @@ def write_list(filename, station_positions, station_name=None, append=False):
...
+    ----------
+    filename :  string
+        should have the extension ".list"
+        If the file is supposed to be used with the 
+        radio_mpi Corsika generator (https://github.com/fedbont94/Horeka/tree/radio_mpi),
+        keep the default filename.
+
+    zenith :  float (in radians)
+        zenith angle of the incoming signal/air-shower direction (0 deg is pointing vertically upwards)
+        Is converted to radians immediately
```
Comment (cg-laser) [inline radiotools/coreas/generate_coreas_sim.py]: what do you mean with this comment? It is not converted to radians, the user input is already in radians.

File: radiotools/coreas/generate_coreas_sim.py [hunk 2]
```diff
@@ -171,67 +171,233 @@ def write_list(filename, station_positions, station_name=None, append=False):
...
+
+    azimuth :  float (in radians)
+        azimuth angle of the incoming signal/air-shower direction (0 deg is North, 90 deg is West)
+        Is converted to radians immediately
```
Comment (cg-laser) [inline radiotools/coreas/generate_coreas_sim.py]: same as above

File: radiotools/coreas/generate_coreas_sim.py [hunk 3]
```diff
@@ -171,67 +171,233 @@ def write_list(filename, station_positions, station_name=None, append=False):
...
+    Auger_CS : bool (default is True)
+        True -> you are providing input in Auger coordinates
+        False -> you are providing input in Corsika coordinates
+
+    inclination :  float (in radians)
+        Inclination of the magnetic field.
+        It describes the angle between the Earth's surface and the magnetic field lines.
+        The default value is given for the Auger site
+
+    Rmin, Rmax, n_rings, arm_orientations : used to calculate the positions of the antennas on the arms of the starshape (!!in m!!)
```
Comment (cg-laser) [inline radiotools/coreas/generate_coreas_sim.py]: you could be more specific here. Typically the user needs to modify Rmin and Rmax and potentially also n_rings.

File: radiotools/coreas/generate_coreas_sim.py [hunk 4]
```diff
@@ -242,21 +408,98 @@ def write_list_star_pattern(filename, zen, az, append=False, obs_level=1564.0, o
...
+                file.write(f"AntennaPosition = {shower_plane_system[i, 0]} {shower_plane_system[i, 1]} {shower_plane_system[i, 2]} {name}\n")
+            
+            print("Saved antenna positions (in vxB_vxvxB coordinates) to file: ", "shower_plane.list")
+
+
+    # return corsika azimuth angle to for automatically generating corsika input files with the right values
+    return corsika_azimuth
+
+
+def get_rmax(X):
```
Comment (cg-laser) [inline radiotools/coreas/generate_coreas_sim.py]: This will be very observation level and to a lesser extent also bandwidth dependent. Can you specify for which settings you estimated the relationship? To improve it for future use, the function could take the observation level as additional input which would allow to add additional parameterizations for other sites later.

**Review by cg-laser — COMMENTED** [review]
Thanks a lot for these improvements. I think it makes these scripts more usable, especially for newcomers. 
I added a few inline comments that should be addressed before merging. I also saw that there is a conflict with the master branch. Can you merge the current master into this branch `git merge master` and resolve the conflicts? 

A general comments about units. We don't enforce it but we are using the default units from NuRadio* https://github.com/nu-radio/NuRadioMC/blob/develop/NuRadioReco/utilities/units.py  (which should be the same as the OFFLINE and GEANT default units). 
Whenever no unit is specified, it should be in these default units. And ideally we always use these default units. 
The only exception is "Grammage" which is defined in g/cm^2 due to historical reasons and because it gets unreasonable values when defined in the NuRadio default units.

File: radiotools/coreas/generate_coreas_sim.py [hunk 5]
```diff
@@ -4,7 +4,9 @@
 import numpy as np
 from radiotools import coordinatesystems
 from radiotools import helper as hp
-from past.builtins import xrange
+from radiotools.atmosphere import models
+import sys
+from miniradiotools.utils.cherenkov_radius import get_cherenkov_radius_model_from_depth
```
Comment (fschlueter) [inline radiotools/coreas/generate_coreas_sim.py]: Is that a new external dependency

File: radiotools/coreas/generate_coreas_sim.py (hunk 1 above)
Comment (lguelzow) [reply radiotools/coreas/generate_coreas_sim.py]: Thanks, this was left over from how I originally implemented this function in Jelena's repository where we used degrees as the input unit for all angles

File: radiotools/coreas/generate_coreas_sim.py (hunk 4 above)
Comment (lguelzow) [reply radiotools/coreas/generate_coreas_sim.py]: The input for the function is already observation level dependent since it uses the grammage at observation level when it's used in the next function . The grammage is calculated with "get_atmospherere" from the models.py file.
In any case, it needs more documentation.

I agree it's not great solution for automatically generating antenna rings. When using this branch to generate simulations recently, I also noticed that the new function "get_starshaped_pattern_radii" that I added has a dependency to a function from @fschlueter's RadioAnalysis framework to estimate the Cherenkov radius of a shower. 
I used these two functions to, independly from geometry, generate antenna rings that go to big axis distances, but have a higher density from the shower core to a little further out than the Cherenkov radius.
I think it's an important functionality to have, but I might have to implement it differently. If I kept it mostly the same, then I need to implement the Cherenkov radius functions into radiotools as well.

File: radiotools/coreas/generate_coreas_sim.py (hunk 5 above)
Comment (lguelzow) [reply radiotools/coreas/generate_coreas_sim.py]: yeah, I noticed it after the original changes and imported it from another another repository in order to use this branch short-term, but I need to fix it so it isn't using anything external. See my last message about the new functions

**Comment by lguelzow** [comment]
Everything should be addressed now. 
I merged master into the branch to resolve the conflicts and I added a new file with @fschlueter's functions to estimate the cherenkov radius (hope that's alright to add), so there's no external ddependencies.

**Review by cg-laser — APPROVED** [review]
I didn't go through all changes again but I trust that you have checked things properly, so let me approve it. Maybe @fschlueter want's to have another look?

File: radiotools/atmosphere/cherenkov_radius.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        a=9.48990456e-01, b=4.48698860e+03,
+        c=1.43097665e+00, d=2.46630811e+06):
+
+    n = atm.get_n(height, n0=n0, model=model)
+    A = a - (b / dist) ** (c) - dist / d
+
+    return A * cherenkov_angle_model(n)
+# # old param
+# def cherenkov_angle_from_density(x, A=0.24905864, B=0.92165234):
+#     return np.deg2rad(A * np.log(x) + B)
```
Comment (fschlueter) [inline radiotools/atmosphere/cherenkov_radius.py]: Is that relevant to keep? I am not sure if that is worth keeping in here. You could also remove the "*model*" from all the function names?

File: radiotools/coreas/generate_coreas_sim.py [hunk 7]
```diff
@@ -241,19 +415,118 @@ def write_list_star_pattern(filename, zen, az, append=False, obs_level=1564.0, o
...
+    if at is None:
+        if atm_model is None:
+            sys.exit("No proper arguments for get_starshaped_pattern_radii")
+
+        at = models.Atmosphere(atm_model)
+
+    # calculate maximum distance of antenna from shower core (in shower plane)
+    # uses rough, hardcoded parametrisation
+    maxX = at.get_atmosphere(zenith, obs_level)
+    rmax = get_rmax(maxX)
```
Comment (fschlueter) [inline radiotools/coreas/generate_coreas_sim.py]: Where is this coming form? Is that a (very) old parameterisation from Christian based on vertical showers? I would tend to define a maximal radius by a multiplicity of cherenkov radii.

File: radiotools/coreas/generate_coreas_sim.py [hunk 8]
```diff
@@ -169,67 +173,237 @@ def write_list(filename, station_positions, station_name=None, append=False):
...
-                            rs=None,
-                            slicing_method=None, slices=[], n_rings=20,
-                            azimuths=np.deg2rad([0, 45, 90, 135, 180, 225, 270, 315]),
-                            gammacut=None):
+def write_list_star_pattern(filename, zenith, azimuth, 
+                            append=False, 
+                            obs_level=1400.0, 
+                            obs_level_corsika=None, 
+                            ground_plane=True,
+                            Auger_CS=True,
```
Comment (fschlueter) [inline radiotools/coreas/generate_coreas_sim.py]: `auger_cs` ?

File: radiotools/atmosphere/cherenkov_radius.py (hunk 6 above)
Comment (fschlueter) [reply radiotools/atmosphere/cherenkov_radius.py]: In my code I added `model` to the function names because I implemented the `revent` functions with the same name, I believe.

Anyway, I think it is a good idea to add them here.

Comment (lguelzow) [reply radiotools/atmosphere/cherenkov_radius.py]: the "old" section is irrelevant and shouldn't be included. I missed it when copying the other functions.
I'll remove the model parts of the function names too, they seem unneccessary here.

File: radiotools/coreas/generate_coreas_sim.py (hunk 7 above)
Comment (lguelzow) [reply radiotools/coreas/generate_coreas_sim.py]: I think so. It's effectively doing that by generating antenna rings that stop about 5 Cherenkov radii out. I just did that with a little trial and error. I agree though that it's more elegant to use the Cherenkov radius directly, especially with the new functions I added. 
I'll make the changes soon.

File: radiotools/coreas/generate_coreas_sim.py (hunk 8 above)
Comment (lguelzow) [reply radiotools/coreas/generate_coreas_sim.py]: That was one of the main reason I started working on this. The old code implicitly assumed that you are using the Auger coordinate system for giving the input parameters and just rotated them to the system CORSIKA uses somewhere in the function. 
The main change I made is make this explicit and documented, and add the option to also use CORSIKA's coordinate system for the input. The Auger_CS option determines the input coordinate system.

Comment (fschlueter) [reply radiotools/coreas/generate_coreas_sim.py]: Sure, may comment was regarding all the capital letters in the variable name.

Comment (lguelzow) [reply radiotools/coreas/generate_coreas_sim.py]: good point, I'll make them lower case

**Comment by lguelzow** [comment]
I addressed Felix' last comments a while ago and I assume I can merge now, but I felt I should with you once again before I do.
Is it ok to merge now?

**Review by fschlueter — APPROVED** [review]
